cumulus/parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+		}
+
+		/// Publish an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::AnnouncementOrigin].
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document to announce.
+		/// - `maybe_expire`: Expiration block of the announcement. If `None`
+		///   [`Config::AnnouncementLifetime`]
+		/// used as a default.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::announce())]
+		pub fn announce(
+			origin: OriginFor<T>,
+			cid: OpaqueCid,
+			maybe_expire: Option<DispatchTime<BlockNumberFor<T>>>,
+		) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let now = frame_system::Pallet::<T>::block_number();
+			let expire_at = maybe_expire
+				.map_or(now.saturating_add(T::AnnouncementLifetime::get()), |e| e.evaluate(now));
+			ensure!(expire_at > now, Error::<T, I>::InvalidExpiration);
+
+			<Announcements<T, I>>::insert(cid.clone(), expire_at);
+			<AnnouncementsCount<T, I>>::try_mutate(|count| -> DispatchResult {
+				*count = count.checked_add(1).ok_or(Overflow)?;
+				Ok(())
+			})?;
+
+			if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |n| n > expire_at) {
+				NextAnnouncementExpireAt::<T, I>::put(expire_at);
+			}
+
+			Self::deposit_event(Event::<T, I>::AnnouncementAnnounced { cid, expire_at });
+			Ok(())
+		}
+
+		/// Remove an announcement.
+		///
+		/// Transaction fee refunded for expired announcements.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::AnnouncementOrigin] or signed for expired
+		///   announcements.
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document to remove.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::remove_announcement())]
+		pub fn remove_announcement(
+			origin: OriginFor<T>,
+			cid: OpaqueCid,
+		) -> DispatchResultWithPostInfo {
+			let maybe_who = match T::AnnouncementOrigin::try_origin(origin) {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			let expire_at = <Announcements<T, I>>::get(cid.clone())
+				.ok_or(Error::<T, I>::MissingAnnouncement)?;
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(maybe_who.is_none() || now >= expire_at, BadOrigin);
+
+			<Announcements<T, I>>::remove(cid.clone());
+			<AnnouncementsCount<T, I>>::mutate(|count| count.saturating_dec());
+
+			Self::deposit_event(Event::<T, I>::AnnouncementRemoved { cid });
+
+			if now >= expire_at {
+				return Ok(Pays::No.into())
+			}
+			Ok(Pays::Yes.into())
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Clean up expired announcements.
+		pub fn cleanup_announcements(now: BlockNumberFor<T>) {
+			if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |next| next > now) {
+				// no expired announcements expected.
+				return
+			}
+			let mut maybe_next: Option<BlockNumberFor<T>> = None;
+			let mut count = 0;
+			<Announcements<T, I>>::translate(|cid, expire_at: BlockNumberFor<T>| {
```

Comment:
This will iterate over all announcements in each call to this function as far as I can say, and remove some. Two risks: 

1. You can mess things up if this unbounded iteration is not weighted well. 
2. Even if weigh it correctly, once the announcement map is large enough, it can never be cleared because of this iteration. 


Instead, you should break this down into smaller pieces. In each `on_idle`, we iterate `n` keys from `Announcements`, not all of them. Fairly similar to the `MBMs`, you want to store the cursor of the last key looked at and continue from there.

Reply:
you are right. 
I am just removing auto cleanup, with the expectation that it will be automated off chain or cleanup manually (remove announcement is free from fee if the announcement expired).